### DISCUSSION TITLE
[defaultLocale] Cannot switch language to defaultLocale

### DISCUSSION THREAD

**Discussion by dimylkas** [amannn/next-intl#1520] [Q&A]
### Description

Config

`export const routing = defineRouting({
  locales: ["en", "es"],
  defaultLocale: "en",
  localePrefix: {
    mode: "as-needed",
  },
});`

- open the site not under "/es" and try to switch to "en". Basically whatever you will set to default, user can't to switch to lang that in under defaultLocale

![image](https://github.com/user-attachments/assets/467caf51-4a4a-4737-8224-11bd32968e30)

Here the swither

```
....
  const currentLocale = useLocale();
  console.log("currentLocale", currentLocale);
  console.log("pathname", pathname);

<Select
...   onChange={(e) => {
          const newLocale = e.target.value as Locale;
          console.log("newLocale", { from: currentLocale, to: newLocale });
          console.log("locale", { locale: newLocale });

          router.replace(pathname, { locale: newLocale });
        }}
...
```

### Verifications

- [X] I've verified that the problem I'm experiencing isn't covered in [the docs](https://next-intl-docs.vercel.app/).
- [X] I've searched for similar, existing issues on [GitHub](https://github.com/amannn/next-intl/issues) and [Stack Overflow](https://stackoverflow.com/search?q=next-intl).
- [X] I've compared my app to [a working example](https://next-intl-docs.vercel.app/examples) to look for differences.

### Mandatory reproduction URL

https://d10cir56sw4n5d.cloudfront.net/de

### Reproduction description

Config

`export const routing = defineRouting({
  locales: ["en", "es"],
  defaultLocale: "en",
  localePrefix: {
    mode: "as-needed",
  },
});`

- open the site not under "/es" and try to switch to "en". Basically whatever you will set to default, user can't to switch to lang that in under defaultLocale

![image](https://github.com/user-attachments/assets/467caf51-4a4a-4737-8224-11bd32968e30)

Here the swither

```
....
  const currentLocale = useLocale();
  console.log("currentLocale", currentLocale);
  console.log("pathname", pathname);

<Select
...   onChange={(e) => {
          const newLocale = e.target.value as Locale;
          console.log("newLocale", { from: currentLocale, to: newLocale });
          console.log("locale", { locale: newLocale });

          router.replace(pathname, { locale: newLocale });
        }}
...
```

 "next-intl": "^3.24.0",

### Expected behaviour

User can switch to defaultLocale

**Comment by dimylkas** [comment]
```

          router.replace(pathname, { locale: newLocale });
          router.refresh(); // this one fix the issue
```

**Reply by bienvenidoY** [reply]
Configurations may conflict
```javascript
// i18n/routing.ts
{ 'localePrefix: 'as-needed''}

//middleware.ts 
{macher: ['/']}
```
I removed macher: ['/'] after fixing the problem

**Comment by amannn** [comment]
Please refer to the examples, which have working locale switchers. This is also covered in the docs.

I'm moving this to a discussion since it's a usage question.

**Reply by herbertxy** [reply]
standalone mode config, and localePrefix: { mode: "as-needed", }, }) config
with  examples https://github.com/amannn/next-intl/tree/main/examples/example-app-router will refer this bug

**Comment by mk0y** [comment]
For me `.refresh` also fixed the issue. I didn't find working example for this use case.
